typed-protocols/src/Network/TypedProtocol/Pipelined.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Network.TypedProtocol.Pipelined where
+
+import Network.TypedProtocol.Core
+
+
+
+data PeerSender ps (pk :: PeerKind) (st :: ps) m a where
```

Comment:
The shortcoming of this definition is that it's not pure functional. If you set `m ~ Identity` then the received transitions can have no influence on the term, they are all ignored. How useful is that? It's going to be borderline useless for anything other than `m ~ IO`, and even then it will be difficult to reason about as we will need to use mutable state in order to react to received transitions.

I have an alternative which admits type safe pipelining in a pure functional style.

Reply:
For our concrete use cases now it fits fine. I agree that it's worth exploring nicer approaches that express the dependencies better for the future. It should be possible to add that in later without affecting too much else.